crates/turbo-tasks-memory/src/edges_set.rs

```diff
@@ -0,0 +1,592 @@
+                    return true;
+                }
+                EdgesEntry::ChildAndCell0(type_id) => {
+                    *self = EdgesEntry::Cell0(*type_id);
+                    return true;
+                }
+                _ => {}
+            },
+            EdgeEntry::Cell(_) => match self {
+                EdgesEntry::Cell0(_) => {
+                    *self = EdgesEntry::Empty;
+                    return true;
+                }
+                EdgesEntry::OutputAndCell0(_) => {
+                    *self = EdgesEntry::Output;
+                    return true;
+                }
+                EdgesEntry::ChildAndCell0(_) => {
+                    *self = EdgesEntry::Child;
+                    return true;
+                }
+                _ => {}
+            },
+            EdgeEntry::Collectibles(_) => {}
+        }
+        if let EdgesEntry::Complex(set) = self {
+            if set.remove(&entry) {
+                self.simplify();
+                return true;
+            }
+        }
+        false
+    }
+
+    fn shrink_to_fit(&mut self) {
+        if let EdgesEntry::Complex(set) = self {
+            set.shrink_to_fit();
+        }
+    }
+
+    fn simplify(&mut self) {
+        if let EdgesEntry::Complex(set) = self {
+            match set.len() {
+                0 => {
+                    *self = EdgesEntry::Empty;
+                }
+                1 => {
+                    let entry = set.iter().next().unwrap();
+                    if matches!(
+                        entry,
+                        EdgeEntry::Output
+                            | EdgeEntry::Child
+                            | EdgeEntry::Cell(CellId { index: 0, .. })
+                    ) {
+                        *self = EdgesEntry::from(*entry);
+                    }
+                }
+                _ => (),
+            }
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct TaskEdgesSet {
+    edges: Box<AutoMap<TaskId, EdgesEntry, BuildHasherDefault<FxHasher>>>,
+}
+
+impl TaskEdgesSet {
+    pub fn new() -> Self {
+        Self {
+            edges: Default::default(),
+        }
+    }
+
+    pub fn insert(&mut self, edge: TaskEdge) -> bool {
+        let (task, edge) = edge.task_and_edge_entry();
+        match self.edges.entry(task) {
+            Entry::Occupied(mut entry) => {
+                let entry = entry.get_mut();
+                entry.insert(edge)
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(EdgesEntry::from(edge));
+                true
+            }
+        }
+    }
+
+    pub fn shrink_to_fit(&mut self) {
+        for entry in self.edges.values_mut() {
+            entry.shrink_to_fit();
+        }
+        self.edges.shrink_to_fit();
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.edges.is_empty()
+    }
+
+    pub fn len(&self) -> usize {
+        self.edges.iter().map(|(_, entry)| entry.len()).sum()
+    }
```

Comment:
I don't really like a `len()` operation that takes O(n) time as it's unintiutive. Callers usually won't feel bad about calling `.len()` inside of a loop because they'll assume it's O(1). Either we should:

- Iteratively keep track of the length as elements are added and removed
- OR rename this to something that makes the cost more obvious, like `compute_len`.

Reply:
It unused, so I removed it